### PR TITLE
Diagnose an unreadable install location in the service itself (#2185)

### PR DESCRIPTION
**Pull request by erikdarlingdata**
Closes the last open half of #2185: the **service** now diagnoses an install location its own account cannot read, instead of leaving the operator with three downstream messages that all point somewhere else.

## What was already shipped, and what was still missing

| | |
|---|---|
| #2186 | decoded `initdb failed (exit code -1073741515)` into `0xC0000135 STATUS_DLL_NOT_FOUND` with the two causes and the two checks |
| #2197 | stopped the missing-credential message advising "start the service once" when a bootstrap had already run and failed |
| #2187 | taught `install-darling.ps1` to refuse a user-profile / UNC / mapped-drive install root before anything is created |

**The gap:** the installer only guards installs that go through it. The README's manual `sc create` path is documented, and registering the exe by hand is exactly what someone does with a zip they just extracted. Those installs still land in the reporter's experience — an empty `Output:`, a bare exit code, then a missing `pg-admin-credential.dpapi` and advice to start the service once, which they had. Nothing named the install directory.

## The change

`DarlingInstallLocation` (new, ~320 lines, mostly the reasoning) classifies the install directory, and `DarlingWorker.ExecuteAsync` calls it as **the first thing it does** — ahead of loading `darling.json`, which an unreadable tree also takes out (`Cannot load configuration` is another message that never names the install directory), and long before the managed-Postgres bootstrap. Order is the point: the cause sits **above** the failure it predicts, in the log an operator reads bottom-up.

One `LogCritical` line. For the reported shape it reads:

> This service is installed somewhere its own account cannot read: `C:\Users\username\Desktop\PerformanceMonitorDarling-3.2.0`. That path is under a user profile, and this service is running as `NT SERVICE\PerformanceMonitor Darling` — which is not you, not SYSTEM, and not Administrators. A directory created under a profile grants access to about those three and nobody else (no `BUILTIN\Users`, no Authenticated Users, no CREATOR OWNER), so the service cannot read its own program files there, and cannot even read back what it writes there itself. The bundled PostgreSQL bootstrap is what fails first: `initdb.exe` dies at exit code -1073741515 (`0xC0000135`, `STATUS_DLL_NOT_FOUND`) in the Windows loader, before it can write a word of output. Everything after that is downstream of THIS and is not the fault — a missing `pg-admin-credential.dpapi`, and advice to start the service once, which starting the service again will not satisfy (#2185). FIX: stop the service, move the install to a machine-scoped local path — `C:\PerformanceMonitorDarling` is the documented one, and a folder created there inherits read + execute for `BUILTIN\Users`, which the service account is a member of — then re-run `install-darling.ps1` from the new location. It updates the service in place; `darling.json` can move with the folder, and the store under `%ProgramData%\PerformanceMonitorDarling` and its credentials stay exactly where they are.

The account is read from `DarlingFileSecurity.ServiceAccountDisplayName`, not assumed: an operator who re-homed the service to a domain account or gMSA (#1802/#1823) must not be told to reason about a virtual account they do not use. UNC and mapped-drive get their own sentences, because a share fails for a different reason (a virtual account reaches the network as the COMPUTER account) than a mapped letter does (it belongs to a logon session a service never joins) — an operator told the wrong one goes and checks the wrong thing.

## Three decisions worth arguing with

**It diagnoses and continues; it does not refuse to start.** Same asymmetry the installer applies to an upgrade, for the same reason — #2187's rejected option 2 is exactly the customization that can make this location work, and stranding a deployment that runs today would be worse than the disease. A service has nobody to ask, so it states the cause and gets on with the start it was asked for. An operator who hand-granted the tree keeps a working service plus a standing note that it is one ACL reset away from breaking, which is true.

**Silent on a console run** (`WindowsServiceHelpers.IsWindowsService()`). An interactive run IS the profile owner, and the README suggests test-driving the exe. That is the one false positive available to this check, and it would be aimed at someone doing exactly what the docs told them to. Non-Windows never reaches it at all, so the compose deployment cannot trip it.

**The decision table is the installer's, not a second one.** Profile root read from Windows rather than a hardcoded `C:\Users`, plus `%USERPROFILE%`; UNC excluding the `\\?\` long-path prefix; a drive letter whose type is network. Two definitions of "a location that cannot work" in two languages would drift, and the one that drifted would be the one nobody was reading.

## Tests

`DarlingServiceInstallLocationTests` — the table, both ways. Twenty-three rows, each carrying WHY, reporting every disagreement in one run. The refusals include the reported Desktop shape, the profile root itself, case / relative-segment / forward-slash spellings, a relocated `ProfilesDirectory`, a redirected profile, a UNC share and a mapped letter. The allowances are the half that decides whether the message is believed when it appears: `C:\PerformanceMonitorDarling`, `C:\ProgramData\...`, `C:\Program Files\...`, `D:\...`, `C:\UsersData` and `C:\Users2` (the boundary a bare `StartsWith` gets wrong), and `\\?\C:\...`. Plus: exactly one critical line, silence for a machine-scoped install, silence on a console run, the three message obligations per verdict, the drive probe never consulted without a letter to ask about, and a source pin that the call site stays Windows-gated and stays ahead of config load / `new DarlingManagedPostgres` / `EnsureRunningAsync`.

`DarlingInstallLocationTests.TheInstallerAndTheService_ReachTheSameVerdict_OverOneTable` — the anti-drift device. Both implementations run over **one** table (`DarlingServiceInstallLocationTests.Cases`), under Windows PowerShell 5.1, and disagreement fails regardless of which side is "right". The PowerShell side is the shipped file: both helpers extracted whole and the two lines that compose them into a verdict lifted **verbatim** out of the script rather than retyped. Only the environment is injected — profile root, operator profile, drive type — which is what the C# side takes as parameters. Both sides are also checked against the table's expectation, so a mutual mistake cannot pass as agreement.

## Verification, and its limits

- `PerformanceMonitor.Darling.Service` and `Darling.Tests` both build clean on macOS (`EnableWindowsTargeting`, 0 warnings).
- `Darling.Tests` targets `net10.0-windows` and **cannot run on macOS**, so a throwaway `net10.0` console harness ran 60 assertions against the real build: the containment boundary, `DriveQualifier`, `Classify`'s precedence and short-circuits, the redirected-profile arm, and the full message text. All pass.
- **Watched red by mutation, both ways.** Dropping the `+ separator` from the containment test reddens the three boundary rows and the table's boundary case; dropping the `\\?\` exclusion reddens the extended-length row. Restored and re-verified green.
- The harness also caught a defect **in itself** rather than in the code: on macOS a Windows-shaped path is *relative*, so `Path.GetFullPath` resolved every one of them under the cwd — which lives under `/Users` — and the profile arm swallowed the UNC and mapped-drive rows before they were reached. Isolating the profile root to `/prof` fixed the harness. Worth saying because it is the reason the POSIX-shaped rows exist: they exercise the algorithm, not Windows normalization.
- **Not verified here:** anything needing Windows. The xUnit suites, the cross-language parity test (it shells out to `powershell.exe`), `DriveInfo.DriveType` against a real mapped letter, `WindowsServiceHelpers.IsWindowsService()` under a real SCM start, and an end-to-end service start from a profile path. CI's `build` and `Darling PostgreSQL tests` jobs are the arbiter for the first two; the last one needs the #2187 verify kit and a physical UAC click.

## Deferred, filed rather than dropped

- **`install-darling.ps1` is unchanged.** Making the C# the single source and having the script call `--check-install-location` is the better end state, and it would change the installer's behavior, which is not this change's to make. Say the word and I will file it; the parity test holds the line in the meantime.
- **A known shared residual, pinned as a table row rather than left to be discovered:** the extended-length spelling of a profile path (`\\?\C:\Users\bob\...`) is not matched, because it does not begin with `C:\Users`. **Both** implementations have it; a fix belongs in both at once rather than in whichever one someone edits first.

Not closing #2185 — the reporter-facing reply is yours, and their last data point (`initdb --version` printing on a machine-scoped path) still says their particular failure was something else. This makes the hypothesis self-reporting for the next person who hits it.

### CODE REVIEW

File: Darling/PerformanceMonitor.Darling.Service/DarlingInstallLocation.cs [hunk 1]
```diff
@@ -0,0 +1,323 @@
...
+            {
+                var parent = Path.GetDirectoryName(Path.TrimEndingDirectorySeparator(Path.GetFullPath(publicProfile)));
+                if (!string.IsNullOrWhiteSpace(parent))
+                {
+                    return parent;
+                }
+            }
+            catch (Exception ex) when (ex is ArgumentException or NotSupportedException or PathTooLongException or IOException)
+            {
+                /* An unusable %PUBLIC% is not a reason to skip the check — fall through to the default. */
```
[the remark this answers is not in the document]
Comment (erikdarlingdata) [reply Darling/PerformanceMonitor.Darling.Service/DarlingInstallLocation.cs]: You're right, and the mechanism is exactly as you describe: `ProfilesDirectory` and `Public` are two independent `REG_EXPAND_SZ` values under one key that merely default to the same tree. My comment asserted a coupling that does not exist. Fixed in 494ade2 by reading `ProfileList\ProfilesDirectory` directly, the same value the installer reads — which removes the divergence rather than documenting it, and is what "share the installer's decision table" was supposed to mean in the first place. No package reference needed: the Registry types are in the shared framework's ref pack for `net10.0`, annotated Windows-only, which the class already is. Same `%SystemDrive%\Users` fallback as the installer, and an unreadable `ProfileList` still does not skip the check.

Your second point is the sharper one and I've closed it executably rather than by softening prose. You're right that both parity tests inject the root — that is what makes the *table* comparable, and it is precisely why neither of them could see this: the lookups themselves were never exercised. `TheInstallerAndTheService_ReadTheProfileRoot_FromTheSamePlace` supplies nothing. It runs the installer's shipped `Get-ProfilesDirectory` and requires `MachineProfileRoot()` to answer the same on whatever box the tests run on. It passes trivially on an unrelocated box, which is the honest limitation — CI is not going to have relocated profiles — but its job is to fail the moment the two stop reading the same thing, which is the failure mode you found.

File: Darling/PerformanceMonitor.Darling.Service/DarlingInstallLocation.cs [hunk 2]
```diff
@@ -0,0 +1,339 @@
...
+                }
+            }
+        }
+        catch (Exception ex) when (ex is SecurityException or UnauthorizedAccessException or IOException or ObjectDisposedException or PlatformNotSupportedException)
+        {
+            /* An unreadable ProfileList is not a reason to skip the check — fall through to the default. */
+        }
+
+        var systemDrive = Environment.GetEnvironmentVariable("SystemDrive");
+        return Path.Combine(string.IsNullOrWhiteSpace(systemDrive) ? "C:" : systemDrive, "Users");
```
[the remark this answers is not in the document]
Comment (erikdarlingdata) [reply Darling/PerformanceMonitor.Darling.Service/DarlingInstallLocation.cs]: Correct, and thank you — that one was worse than it looks. `Path.Combine` treating a trailing *volume* separator the same as a trailing directory separator gives `C:Users`, `Path.GetFullPath` then anchors it to the process's current directory on the volume rather than the volume root, and for a service that is not `C:\`. So the profile check would have silently stopped matching on exactly the box the fallback exists for. And you're right that it was a second silent divergence from the installer, whose `Join-Path 'C:' 'Users'` does the right thing.

Fixed in 2802576. The composition is now its own pure function, `ProfileRootForSystemDrive`, for the reason you identified: the registry read succeeds on every box CI runs on, so the fallback branch was unreachable from a test, which is why it shipped. It's pinned by a table — bare `C:`, `C:\` with a separator already present, `D:`, unset, blank, null — asserting `drive + separator + Users` and specifically **not** bare concatenation, plus an assertion that whichever branch the shipped lookup takes, `MachineProfileRoot()` returns a fully-qualified path.

Worth recording since it explains why my local verification was blind to it: the macOS harness could not have caught this. `Path.Combine`'s volume-separator special case is Windows-only behavior, and `Path.Combine("C:", "Users")` already yields `C:/Users` there. Two findings in a row now that only Windows semantics expose — the harness's scope is the algorithm, not the platform, and I should have said so more narrowly.

File: Darling/PerformanceMonitor.Darling.Service/DarlingInstallLocation.cs [hunk 3]
```diff
@@ -0,0 +1,362 @@
...
+    /// instead of documenting it (review catch on #2185).</para>
+    ///
+    /// <para>Falls back to <c>%SystemDrive%\Users</c>, which is what the installer falls back to. An
+    /// unreadable <c>ProfileList</c> is not a reason to skip the check.</para>
+    /// </summary>
+    internal static string MachineProfileRoot()
+    {
+        try
+        {
+            using var profileList = Registry.LocalMachine.OpenSubKey(
```
[the remark this answers is not in the document]
Comment (erikdarlingdata) [reply Darling/PerformanceMonitor.Darling.Service/DarlingInstallLocation.cs:315]: It does compile, and I should show the evidence rather than assert it: `Darling Linux build`, `build` (Windows) and `Darling PostgreSQL tests` all passed on 2802576, the commit that introduced `using Microsoft.Win32;`, and it builds clean on macOS too (`-p:EnableWindowsTargeting=true`, 0 warnings). So no package reference is needed.

The reason is that `Microsoft.Win32.Registry` is not in the same position as `System.IO.FileSystem.AccessControl`. The `RegistryKey` types live in the shared framework's reference assemblies for plain `net10.0` — annotated `[SupportedOSPlatform("windows")]`, which is why the analyzer is satisfied by the `[SupportedOSPlatform("windows")]` already on this class and why there is no CA1416 — whereas the AccessControl *extension methods* genuinely are not, hence that package and that comment. Different situation, so the csproj comment does not generalize the way it reads like it should.

Runtime is covered separately and twice: the call site is inside `if (OperatingSystem.IsWindows())`, and `PlatformNotSupportedException` is in the catch list on the registry read regardless, so a non-Windows host falls to the `%SystemDrive%\\Users` default rather than throwing.

File: Darling/PerformanceMonitor.Darling.Service/DarlingInstallLocation.cs [hunk 4]
```diff
@@ -0,0 +1,362 @@
...
+                    .Append(" — a drive letter belongs to the logon session that mapped it, which a service does not share and cannot see at all, and a virtual service account reaches the network as the COMPUTER account rather than as you.");
+                break;
+
+            default:
+                /* Describe is only reached with a real verdict; Report returns on None. Kept explicit rather
+                   than throwing, because a diagnostic must never be the thing that fails. */
+                break;
+        }
+
+        message.Append(" The bundled PostgreSQL bootstrap is what fails first: initdb.exe dies at exit code -1073741515 (0xC0000135, STATUS_DLL_NOT_FOUND) in the Windows loader, before it can write a word of output. Everything after that is downstream of THIS and is not the fault — a missing pg-admin-credential.dpapi, and advice to start the service once, which starting the service again will not satisfy (#2185).")
```
[the remark this answers is not in the document]
Comment (erikdarlingdata) [reply Darling/PerformanceMonitor.Darling.Service/DarlingInstallLocation.cs]: Fair, and fixed in 3eecc5e. You're right on both halves: the initdb chain only exists when `postgres.managed` is true, and this message is deliberately built *before* `darling.json` is read — because an unreadable tree takes the config out too — so it cannot know which mode is configured. Narrating the managed chain unconditionally would send a bring-your-own operator looking for an initdb that never runs, which spends exactly the credibility the message exists to have.

I kept the ordering and named both instead, since deferring the message until after config load would cost the property the whole change is for. Two sentences now: the managed default's chain, labelled as the default's, and then the one a BYO operator actually sees — `"Cannot load configuration"` from a `darling.json` sitting in the same unreadable folder. Your point 1 was the sharper one there: that alternative was missing entirely even though the call site's own comment cites it as a reason to run the check early. `Describe_NamesThePathTheReasonAndTheRemedy` now requires both modes and the config-load string, for every verdict.

**Comment by erikdarlingdata** [comment]
Reviewed the diff, merging. The risk with any new startup check is that it breaks working installs, so that is what I verified rather than the message text:

- Gated on `OperatingSystem.IsWindows()` at the call site **and** an early return unless `IsWindowsService()` — so a console test-drive from a Desktop folder, which the README actively suggests, stays silent. That was the one available false positive and it is closed.
- It **diagnoses and continues** rather than refusing to start. Right call for the reason you gave: the installer asks rather than refuses on an upgrade because an operator may have hand-granted the tree, and a service has nobody to ask. A refusal here would turn a misdiagnosed-but-working install into a dead one.
- Running ahead of `DarlingConfig.Load()` rather than just ahead of the Postgres bootstrap is the detail that makes it actually work — an unreadable tree takes out `darling.json` too, so `Cannot load configuration` is a second misdirecting message, and the cause now sits above every failure it predicts.

The parity test running both the PowerShell installer's composition and the C# implementation over one shared table is a better answer than refactoring one into the other under time pressure, and #2348 correctly carries the extended-length residual that needs both halves moved together.

Also noted: you pushed back on the review's `Microsoft.Win32.Registry` claim with three passing builds instead of complying. That is the right response to a review finding that is simply wrong.
